### PR TITLE
Ban use of defparam

### LINKED ISSUE #12: Ban use of defparam
Our to-be-published style guide explicitly bans the use of the `defparam` construct.
This feature request aims to enforce just that.

**Assignment:**
Implement this check as a rule inside here:
https://github.com/google/verible/tree/master/verilog/analysis/checkers

Include unit-tests, positive and negative.

Add a TODO to enable by default here:
https://github.com/google/verible/blob/master/verilog/analysis/default_rules.h

Add an integration test here:
https://github.com/google/verible/blob/master/verilog/tools/lint/BUILD

**Suggestion:**
Use a syntax tree matcher, starts with 'Nodek', defined in https://github.com/google/verible/blob/master/verilog/CST/verilog_matchers.h

Questions? ask in verible-dev@groups.google.com!
https://groups.google.com/forum/#!forum/verible-dev

### PR DESCRIPTION
**Pull request by tgorochowik**
Fixes #12

### CODE REVIEW

File: verilog/analysis/checkers/defparam_rule.cc [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+namespace verilog {
+namespace analysis {
+
+using verible::GetStyleGuideCitation;
+
+// Register the lint rule
+VERILOG_REGISTER_LINT_RULE(DefparamRule);
+
+absl::string_view DefparamRule::Name() { return "defparam"; }
```
Comment (fangism) [inline verilog/analysis/checkers/defparam_rule.cc]: Let's register this ::Name() as "forbid-defparam".

File: verilog/analysis/checkers/defparam_rule.cc [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+  return absl::StrCat(
+      "Do not use defparam. See:",
+      GetStyleGuideCitation(kTopic), ".");
+}
+
+void DefparamRule::HandleSymbol(
+    const verible::Symbol& symbol, const verible::SyntaxTreeContext& context) {
+  verible::matcher::BoundSymbolManager manager;
+  if (matcher_.Matches(symbol, &manager)) {
+    violations_.push_back(verible::LintViolation(symbol, kMessage, context));
```
Comment (fangism) [inline verilog/analysis/checkers/defparam_rule.cc]: I recommend narrowing the range to that of the 'defparam' token itself, which should be in child position [0].

File: verilog/analysis/checkers/defparam_rule_test.cc [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+namespace {
+
+using verible::LintTestCase;
+using verible::RunLintTestCases;
+
+TEST(DefparamRuleTests, Various) {
+  constexpr int kToken = SymbolIdentifier;
+  const std::initializer_list<LintTestCase> kTestCases = {
+      {""},
+      {"module parametrized; parameter int MY_PARAM = 0; endmodule; "
```
Comment (fangism) [inline verilog/analysis/checkers/defparam_rule_test.cc]: 'parameterized' (spelling)

File: verilog/analysis/checkers/defparam_rule_test.cc [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+namespace verilog {
+namespace analysis {
+namespace {
+
+using verible::LintTestCase;
+using verible::RunLintTestCases;
+
+TEST(DefparamRuleTests, Various) {
+  constexpr int kToken = SymbolIdentifier;
```
Comment (fangism) [inline verilog/analysis/checkers/defparam_rule_test.cc]: With my suggestion to narrow down to the keyword, this should be the token enum TK_defparam.

File: verilog/analysis/checkers/defparam_rule_test.cc [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+  const std::initializer_list<LintTestCase> kTestCases = {
+      {""},
+      {"module parametrized; parameter int MY_PARAM = 0; endmodule; "
+       "module foo; parametrized p0(); endmodule"},
+
+      {"module parametrized; parameter int MY_PARAM = 0; endmodule; "
+       "module foo; ",
+       "defparam ", {kToken, "p0"}, ".MY_PARAM = 1; "
```
Comment (fangism) [inline verilog/analysis/checkers/defparam_rule_test.cc]: With suggested change, this should expect to point to {kToken, "defparam"}.

File: verilog/analysis/default_rules.h [hunk 6]
```diff
@@ -52,6 +52,7 @@ constexpr const char* kDefaultRuleSet[] = {
     // TODO(b/117554159): "case-missing-default",
     // TODO(b/117131903): "proper-parameter-declaration",
     // TODO(b/131637160): "signal-name-style",
+    // TODO(b/138351837): "defparam",
```
Comment (fangism) [inline verilog/analysis/default_rules.h]: same as before "forbid-defparam"

File: verilog/analysis/checkers/defparam_rule.cc (hunk 2 above)
Comment (tgorochowik) [reply verilog/analysis/checkers/defparam_rule.cc]: Thank you for the review!

Could you please elaborate on that? As far as I can see some other checkers try to use the BoundMap from the manager to extract more detailed info, but it is empty in this case. Whose child node should be used here, should it be extracted from the context? And what is the proper way to access it?

Comment (tgorochowik) [reply verilog/analysis/checkers/defparam_rule.cc]: `violations_.push_back(verible::LintViolation( *(context.top().children()[0].get()), kMessage, context));`

Seems to work, is that what you meant?

**Comment by tgorochowik** [comment]
@googlebot I signed it!

File: verilog/analysis/checkers/defparam_rule.cc (hunk 1 above)
Comment (tgorochowik) [reply verilog/analysis/checkers/defparam_rule.cc]: Sure! updated

File: verilog/analysis/checkers/defparam_rule.cc (hunk 2 above)
Comment (tgorochowik) [reply verilog/analysis/checkers/defparam_rule.cc]: Changed the implementation to what I pasted above (+ some sanity checks), please let me know if that's what you had in mind

File: verilog/analysis/checkers/defparam_rule_test.cc (hunk 3 above)
Comment (tgorochowik) [reply verilog/analysis/checkers/defparam_rule_test.cc]: Right, thanks!

File: verilog/analysis/checkers/defparam_rule_test.cc (hunk 4 above)
Comment (tgorochowik) [reply verilog/analysis/checkers/defparam_rule_test.cc]: Updated

File: verilog/analysis/checkers/defparam_rule_test.cc (hunk 5 above)
Comment (tgorochowik) [reply verilog/analysis/checkers/defparam_rule_test.cc]: Updated

File: verilog/analysis/default_rules.h (hunk 6 above)
Comment (tgorochowik) [reply verilog/analysis/default_rules.h]: Updated

**Comment by tgorochowik** [comment]
@fangism thank you very much for your review! I've added the integration tests and addressed your inline comments.

One additional question: should the class (and files) for the rule also be renamed to ForbidDefparamRule?

File: verilog/tools/lint/testdata/defparam_usage.sv [hunk 7]
```diff
@@ -0,0 +1,7 @@
+module defparam_usage #(parameter int MY_PARAM = 0);
+endmodule;
+
+module foo;
+  defparam p0.MY_PARAM = 1;
```
Comment (fangism) [inline verilog/tools/lint/testdata/defparam_usage.sv]: Just add a comment marking this line as the one expected to trigger a the lint rule "forbid-defparam".

File: verilog/analysis/checkers/defparam_rule.h [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+  static std::string GetDescription(DescriptionType);
+
+  void HandleSymbol(const verible::Symbol& symbol,
+                    const verible::SyntaxTreeContext& context) override;
+
+  verible::LintRuleStatus Report() const override;
+
+ private:
+  // Matches the defparam construct.
+  const verible::matcher::Matcher matcher_ = NodekDefParamAssignList();
```
Comment (fangism) [inline verilog/analysis/checkers/defparam_rule.h]: I overlooked this before, I would recommend matching on the CST node created here:

https://github.com/google/verible/blob/f74e900dbdd40c48fcaf1ff68db0c7f5bd342658/verilog/parser/verilog.y#L5286

so the node tagged `NodeEnum::kParameterOverride`.

File: verilog/analysis/checkers/defparam_rule.cc [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+      "Do not use defparam. See:",
+      GetStyleGuideCitation(kTopic), ".");
+}
+
+void DefparamRule::HandleSymbol(
+    const verible::Symbol& symbol, const verible::SyntaxTreeContext& context) {
+  verible::matcher::BoundSymbolManager manager;
+  if (matcher_.Matches(symbol, &manager)) {
+    if (!context.empty() && !context.top().children().empty()) {
+      auto defparam = context.top().children()[0].get();
```
Comment (fangism) [inline verilog/analysis/checkers/defparam_rule.cc]: (See comment in .h file first on suggested node matcher.)
Take a look at GetSubtreeAsLeaf:
https://github.com/google/verible/blob/f74e900dbdd40c48fcaf1ff68db0c7f5bd342658/common/text/tree_utils.h#L102
With the update to the matcher, symbol will match at the node tagged `kParameterOverride`, whose child[0] will point to the `defparam` keyword leaf.  Then you don't need to manipulate the `context`.

```
const auto& defparam_token = GetSubtreeAsLeaf(symbol, NodeEnum::kParameterOverride, 0).get();
CHECK_EQ(defparam_token.token_enum, TK_defparam);
violations_.push_back(verible::LintViolation(defparam_token, kMessage, context));
```

File: verilog/analysis/checkers/defparam_rule_test.cc [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+TEST(DefparamRuleTests, Various) {
+  constexpr int kToken = TK_defparam;
+  const std::initializer_list<LintTestCase> kTestCases = {
+      {""},
+      {"module parameterized #(parameter int MY_PARAM = 0); endmodule; "
+       "module foo; parameterized p0(); endmodule"},
+
+      {"module parameterized #(parameter int MY_PARAM = 0); endmodule; "
+       "module foo; ",
+       {kToken, "defparam"}, " p0.MY_PARAM = 1; "
```
Comment (fangism) [inline verilog/analysis/checkers/defparam_rule_test.cc]: With the above changes, I expect this test will now match the expected `{kToken, "defparam"}` violation.

File: verilog/analysis/checkers/defparam_rule_test.cc [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+       "parameterized p0(); endmodule"},
+  };
```
Comment (fangism) [inline verilog/analysis/checkers/defparam_rule_test.cc]: Some suggestions for additional test cases:
* a test case with two violations (could be in the same module, or in different modules)
* a test case where the "defparam" is a substring of a (non-keyword) identifier, e.g. `wire not_defparam;`
* a test case where "defparam" is buried inside a /* comment */ or // comment.

File: verilog/tools/lint/testdata/defparam_usage.sv (hunk 7 above)
Comment (tgorochowik) [reply verilog/tools/lint/testdata/defparam_usage.sv]: Sure! Added a comment above, like other tests do

File: verilog/analysis/checkers/defparam_rule.h (hunk 8 above)
Comment (tgorochowik) [reply verilog/analysis/checkers/defparam_rule.h]: Ok, sure, that makes sense

File: verilog/analysis/checkers/defparam_rule.cc (hunk 9 above)
Comment (tgorochowik) [reply verilog/analysis/checkers/defparam_rule.cc]: Right, that looks much cleaner this way, thanks for the detailed explanation!

File: verilog/analysis/checkers/defparam_rule_test.cc (hunk 11 above)
Comment (tgorochowik) [reply verilog/analysis/checkers/defparam_rule_test.cc]: Sure, added the suggested test cases

**Comment by tgorochowik** [comment]
@fangism Thank you again for the review. I just pushed an update which renames the files and addresses all your comments
